sdk/include/opentelemetry/sdk/trace/sampler.h

```diff
@@ -0,0 +1,91 @@
+#pragma once
+
+#include "opentelemetry/common/attribute_value.h"
+#include "opentelemetry/nostd/span.h"
+#include "opentelemetry/nostd/string_view.h"
+#include "opentelemetry/sdk/common/atomic_shared_ptr.h"
+#include "opentelemetry/trace/span.h"
+#include "opentelemetry/trace/trace_id.h"
+#include "opentelemetry/trace/tracer.h"
+#include "opentelemetry/version.h"
+
+#include <map>
+#include <memory>
+#include <string>
+
+OPENTELEMETRY_BEGIN_NAMESPACE
+namespace sdk
+{
+namespace trace
+{
+namespace trace_api = opentelemetry::trace;
+/**
+ * NOT_RECORD - IsRecording() == false,
+ *    span will not be recorded and all events and attributes will be dropped.
+ * RECORD - IsRecording() == true, but Sampled flag MUST NOT be set.
+ * RECORD_AND_SAMPLED - IsRecording() == true AND Sampled flag` MUST be set.
+ */
+enum class Decision
+{
+  NOT_RECORD,
+  RECORD,
+  RECORD_AND_SAMPLE
+};
+
+/**
+ * A placeholder for common::AttributeKeyValue
+ * A key/value pair that can be used to set attributes.
+ */
+struct AttributeKeyValue {
+    nostd::string_view key;
+    common::AttributeValue value;
+};
+
+struct SamplingResult
+{
+  Decision decision;
+  // TODO: Change AttributeKeyValue to common::AttributeKeyValue
+  std::unique_ptr<nostd::span<AttributeKeyValue>> attributes;
+};
+
+class Sampler
+{
+public:
+  // Placeholder
+  class SpanContext
+  {};
+  virtual ~Sampler() = default;
+  /**
+   * Called during Span creation to make a sampling decision.
+   *
+   * @param parent_context TODO: a shared pointer of the SpanContext of a parent Span.
```

Comment:
What is the reasoning for shared pointer/const pointer?  

How you have it implemented now, it seems like const reference would be more appropriate?

Reply:
It used const pointer because a `parent_context` can be NULL when a span has no parent.
I used const reference when the parameter is not null-able such as `attributes`

Reply:
I don't know if this matches the style guide for this project but I am always a fan of std::optional over null pointers.

the type would be `std::optional<const SpanContext&>`.  Let me know what you think.

Reply:
I haven't thought about that. I think it looks more elegant. I will bring this up during today's SIG meetings to check if there is a specific style requirement.

Reply:
From Gitter chat:
> std::optional is only available in C++17, not for previous standards. We would have to add our own implementation to nostd to have it available for C+11 too. std::optional is nice, but in my opinion not worth the effort adding and maintaining it in nostd.